EIPS/eip-3779.md

```diff
@@ -0,0 +1,180 @@
+
+### Validation
+
+The following is a pseudo-Go implementation of an algorithm for enforcing program validity.  It is a symbolic execution that recursively traverses the bytecode, following its control flow and stack use and checking for violations of the rules above. 
+
+This algorithm runs in time equal to `O(vertices + edges)` in the program's control-flow graph, where vertices represent control-flow instructions and the edges represent basic blocks  thus the algorithm takes time proportional to the size of the bytecode.
+
+For simplicity's sake:
+* We assume an `advance_pc()` routine to skip immediate data.
+* We assume `add_items()` and `remove_items()` functions that push or pop items and return the effect of each instruction on the stack pointer.  `add_items()` will PUSH the value of constants - all other stack items are zeroed.
+```
+   var bytecode [code_len]byte
+   var stack_depth [code_len]unsigned
+   var stack[1024]unsigned
+   var SP := 0
+
+   func validate(PC :=0) boolean {
+      // traverse code sequentially
+      // recurse for subroutines and conditional jumps
+      while true {
+         instruction = bytecode[PC]
+         if is_invalid(instruction) {
+            return false;
+         }
+
+         // stack depth must be the same
+         if SP != stack_depth[PC] {
+               return false
+         }
+         // if stack depth non-zero we have been here before 
+         // return true to break cycle in control flow graph
+         if stack_depth[PC] != 0 { 
+            return true
+         }
+         stack_depth[PC] = SP
+
+          if instruction == JUMP {
+
+             // check for valid destination
+             jumpdest = stack[SP]
+             if bytecode[jumpdest] != JUMPDEST {
+                 return false
+             }
+
+             // reset PC to destination of jump 
+             PC += jumpdest
+             continue
+         }
+         if instruction == JUMPI {
+
+             // check for valid destination
+             jumpdest = stack[SP]
+             if bytecode[jumpdest] != JUMPDEST {
+                 return false
+             }
+
+             // reset PC to destination of jump 
+             PC += jumpdest
+
+             // recurse to jump to code to validate 
+             if !validate(stack[SP])) {
+                 return false
+             }
+             continue 
+         }
+         
+        // effect of instruction on stack
+         SP -= remove_items(instruction)
+         SP += add_items(instruction)
+         if SP < 0 || 1024 < SP {
+             return false
+         }
+
+         // successful validation of path
+         if     PC > code_len ||
+                instruction == STOP || 
+                instruction == RETURN ||
+                instruction == SUICIDE {
+             return true
+         }
+
+         // advance PC according to instruction
+         PC = advance_pc(PC, instruction)
+      }    
+   }
+```
+
+## Rationale
```

Comment:
Consider moving some of the content from the specification into the rationale.  Anything that explains *why* a particular thing is made should be in the rationale, while the specification should be strictly the things required to implement this change.

I like to think of the specification like a recipe book, where the author doesn't explain *why* you are adding cumin to the dish, they just tell you exactly how much cumin to add.